### DISCUSSION TITLE
Quality of RESTEasy reactive

### DISCUSSION THREAD

**Discussion by sithmein** [quarkusio/quarkus#25265] [Community]
Quarkus 2.8 suggests that users should switch to RESTEasy reactive in favour of the classic non-reactive version. However, we have already come across two bugs (issues #25258 and #25263) (where the reactive version violates the JAX-RS specification) and one very [annoying behaviour](https://stackoverflow.com/questions/72046345/errors-after-switching-to-quarkus-resteasy-reactive?noredirect=1#comment127317757_72046345) (accidentially mixing reactive and non-reactive dependencies) just within one day of trying to switch to reactive RESTEasy. This makes me doubt that the reactive version is actually production ready. But then again bugs in the classic version are not fixed any more, only in the reactive version (issues #24738 and #24736). Maybe you should reconsider this decision and apply some serious testing on the reactive version first, before you declare it the new standard. Just my two cents (and yes I'm a bit angry right now because I lost a whole day finding out why things suddenly don't work any more).

**Comment by gsmet** [comment]
> and yes I'm a bit angry right now because I lost a whole day finding out why things suddenly don't work any more

And so you think it's a good idea to criticize months of work by other people without taking a step back?

RESTEasy Reactive passes the JAX-RS TCK (except some very weird behaviors) so if things are not in line with the spec, I would say the main culprit is an incomplete TCK. Implementing a spec is extremely hard and that's why TCKs exist. Now even the perfect TCK cannot cover all the corner cases out there, especially since it's implementation agnostic and corner cases are often tied to implementation details.

Now, RESTEasy Reactive has been around for a while and is used in production so its infrastructure works and is something we want for Quarkus. Your issues will be fixed and thanks for reporting them. They will be fixed even faster if you provide full Maven reproducers.

Or you can also help fix them given it's all Open Source.

**Comment by sithmein** [comment]
You have to consider how users of Quarkus are feeling. We report a bug and it's fixed (quite fast, which I really appreciate). But in order to get the fix, we have to switch to a completely different JAX-RS implementation which introduces new bugs. And so we are stuck. I assume these new bugs will be fixed soonish again but very likely it requires us updating to 2.9 which adds new features which in turn can bring with it new bugs (that's the nature of new features). This can go on and on forever but in the end we have to finish the project at some point in time.
What I believe is missing here is a kind of LTS version of Quarkus that does not add new features (personally we don't need any of the new features added recently) but fixes bugs only.

And I disagree that the two last bugs we reported are edge-cases. Using _resource-id_ as path parameter (which we do in our code) isn't out of the ordinary at all. Also specifying no content type in the response is even standard I would say, otherwise you are not able to deliver more than one representation of a resource. Given that nobody has come across these issues so far makes me believe (you may have a different opinion and this is absolutely fine) that the reactive version is less used and mature than the classic version. And then declaring it the new standard and even stopping support for the classic version is kind of a strange decision from my perspective.

About contributing to an open source project: our company maintains a huge open source project but in the end we have to make some money, too, in order to pay our salaries. I'm totally fine with creating minimal reproducers for bugs but diving into the Quarkus code would require me days into order to get familiar and meanwhile our own project is on hold.

**Reply by geoand** [reply]
It's totally understandable that you think the issues you ran into are standard issues. But if the TCK doesn't test for them and none of our other users have run into them, then I would argue that they are edge cases.

**Comment by gsmet** [comment]
> You have to consider how users of Quarkus are feeling.

Well, we certainly have. We have been listening to users since day 1 and are extremely open about everything.

Now, maybe you should consider how Quarkus developers are feeling. I have not worked on RESTEasy Reactive myself so I'm not biased on this and speaking very calmly but you are a Quarkus user, you benefit from our work and well there is not even one positive note in your message, which is your first message to us.

I'm not fishing for compliments or whatever but, really, your one day of frustration is probably nothing compared to what Quarkus brought to you and the productivity you gained using it and starting with a positive note would have been better. Maybe?

Especially, since we haven't removed RESTEasy Classic and have explicitly stated it wasn't going away so you migrating your app now is a choice of yours.
And it's definitely interesting for us to have users doing that and reporting the issues they meet so thanks for that.

> but very likely it requires us updating to 2.9 which adds new features which in turn can bring with it new bugs (that's the nature of new features)

That's the nature of new features and that's also the nature of fixes. Some fixes will introduce regressions too.

My personal opinion is that we have a nice balance and we are innovating while making micro releases that fix bugs and that overall the Quarkus experience is nice - but not perfect, that's for sure.

> What I believe is missing here is a kind of LTS version of Quarkus that does not add new features

It's not missing. It's called the Red Hat Build of Quarkus. If you want LTS, that's the way to go.

And note that we don't use it as an excuse for issues in Quarkus: we are doing a LOT of micro releases, we fix bugs and we are all working hard on making Quarkus better, stable and as bug free as possible.

A lot of fixes are shipped in less than a week.

> And I disagree that the two last bugs we reported are edge-cases.

I was talking about edge cases of the TCK i.e. things present in the spec but not covered by the TCK itself. I agree that the issues you reported are quite common and needs fixing.

Now, RESTEasy Reactive has been around for a while, is tested by the TCK, by our test suite and is used by some users. At some point, we have to make it more mainstream and sure people will have issues at first but that's how things are.

The exact same way that when we are releasing CRs for testing, a lot of people are not testing it and waiting for the Final to test their app.

At some point, we have to take a leap of faith.

> Given that nobody has come across these issues so far makes me believe (you may have a different opinion and this is absolutely fine) that the reactive version is less used and mature than the classic version. And then declaring it the new standard and even stopping support for the classic version is kind of a strange decision from my perspective.

You're certainly entitled to have an opinion. We strongly believe RESTEasy Reactive is the way to go and we are committed to making it rock solid.

Now my guess is that it will be far easier if people are collaborating with us and contributing to the community - be it by reporting issues with reproducers or contributing code - rather than with this sort of post.

**Reply by idc77** [reply]
It's not like you're not getting paid for your work

**Reply by gsmet** [reply]
@idc77 so when you go to the bakery, you don't say thank you because the person in front of you is paid? That's one way of doing it. It certainly makes the world a little bit worse.

And btw, while I'm paid to work on Quarkus - but I also work on it outside of my working hours -, that's not the case of all the members in the community. So, yeah, treating people and their work with respect should be standard. And that doesn't prevent reporting issues, asking questions, and talk openly about things - critics included.

**Reply by geoand** [reply]
Yes we are extremely fortunate to be paid by Red Hat for developing Quarkus. But being paid to develop Open Source software doesn't imply what all of us are doing, which is going **way** above and beyond to develop new features, fix user reported bugs and interact with the community in what is generally regarded as an astounding rate.

**Comment by sithmein** [comment]
In found the [next bug](https://github.com/quarkusio/quarkus/issues/25283) in reactive RESTEasy. And this is likely not the last one because we have dozens of failing tests since we switched to reactive RESTEasy. Stay tuned...

**Reply by geoand** [reply]
Thanks for reporting these. The feedback is extremely valuable.

**Comment by sithmein** [comment]
I had a look at the _Red Hat Build of Quarkus_. The latest version is 2.2.5.Final which looks very similar to 2.2.5.Final from the "community" version. In what way is this an LTS version? Does it contain more bug fixes than the community version?

**Reply by geoand** [reply]
The next `Red Hat Build of Quarkus` will be based on Quarkus 2.7

**Comment by sberyozkin** [comment]
@sithmein Are you a Red Hat customer ? If yes then please get in touch via the corresponding channels and raise whatever concerns you have.

I have to admit, it has been awhile since I saw so much negativity in an open-source project. Resteasy Reactive is an innovation at its best, it might take some time for those users who depend on most of the RestEasy Classic features to migrate. Many of Quarkus users are spending their own time to help out with fixing some of the related issues. And as has already been said here, I'd also encourage you to go ahead and contribute, to Resteasy Reactive, or to some existing Resteasy classic issues if you prefer. Please help but not demand the fixes.

**Comment by sithmein** [comment]
No, we are not Red Hat customers. But it was pointed out that the _Red Hat Build of Quarkus_ is the LTS version of Quarkus but from what I can see from what's publicly available there is no difference to the community version. This is why I was asking about the actual differences.
And you didn't quite get the point of my complaints. I'm not complaining about RESTEasy reactive per-se I'm questioning it's matureness compared to classic RESTEasy. And  the decision to make it the new default and stop maintaining the classic version (meaning it does not receive any bug fixes any more). All this based on the the bug reports I have already filed.
As you can read above, we are forced to switch to the reactive version in order to get the fix for a bug that we reported (which took my colleague more than two days to track down). But that change revealed several other bugs in the reactive version that prevents us from switching (tracking down these bugs and providing minimal reproducers took us another two days). So as I mentioned before we are currently stuck. And we invested quite some time for minimal reproducers so I guess this qualifies as helping and contributing the project.

**Reply by sberyozkin** [reply]
@sithmein I believe `RestEasy Classic` will be around for a while, fixes may still be applied to it, but please appreciate that `RestEasy Reactive` authors may not have enough time to support both in equal measure. There will be some teething problems for large (classic) JAX-RS projects. 

> And we invested quite some time for minimal reproducers so I guess this qualifies as helping and contributing the project.

I agree. Since your team is looking after a large open source project, I think you can agree that with everyone in the team looking after many issues it may not be possible to react immediately to every user issue. This is when an open source contribution mode helps to accelerate things

**Reply by geoand** [reply]
> And we invested quite some time for minimal reproducers so I guess this qualifies as helping and contributing the project

I can't thank you enough for these. This is definitely open source collaboration at its finest

**Comment by sithmein** [comment]
I really don't want to rant but this is the [fifth broken functionality](https://github.com/quarkusio/quarkus/issues/25312) in reactive RESTEasy that we found.

**Reply by geoand** [reply]
Your issues have almost all been fixed in almost no time, so I would say things are in pretty good state.

**Comment by geoand** [comment]
I'm truly sorry to hear you are having issues.

However I am going to say categorically that we are **not** going to be switching back to RESTEasy Classic as our main focus for our REST stack for reasons that have already been stated in multiple blog posts.

Also it's very important for everyone here to understand that RESTEasy Reactive has been around for over 10 Quarkus versions, so folks have had ample time to try it out and report issues.
Moreover, we have had multiple open source users as Red Hat internal folks switch to RESTEeasy Reactive over the past year and this feedback has been invaluable in flushing out bugs that have not been caught by the TCK.

**Comment by sithmein** [comment]
Some of the issues I have reported have been fixed in 2.8.3. However, there are still some open problems remaining. One is about routing requests to the correct resource method. There are no problem with classic RESTEasy but about 20 methods cannot be found with reactive RESTEasy. The method does not get invoked and a 404 is being returned. This makes me believe there is a bug somewhere in the routing code. Where would I need to start looking so that I can find out the reason? I wasn't able to find a pattern by just looking at our code.

**Reply by geoand** [reply]
There is a class called ClassRoutingHandler (IIRC) that performs the routing.

If you put together a sample that works in RESTEasy Classic and not Reactive, I'll gladly take a look

**Reply by sithmein** [reply]
https://github.com/quarkusio/quarkus/issues/25462
